test/unit/org/apache/cassandra/cql3/validation/operations/InsertTest.java

```diff
@@ -247,4 +252,42 @@ public void testCKInsertWithValueOver64K() throws Throwable
         assertInvalidThrow(InvalidRequestException.class,
                            "INSERT INTO %s (a, b) VALUES ('foo', ?)", new String(TOO_BIG.array()));
     }
+
+    @Test
+    public void testInsertingMapDataWithParameterizedQueriesIsKeyOrderIndependent() throws Throwable
+    {
+        UUID uuid1 = UUIDs.timeBased();
+        UUID uuid2 = UUIDs.timeBased();
+        createTable("CREATE TABLE %s (k text, c frozen<map<timeuuid, text>>, PRIMARY KEY (k, c));");
+        // We use `executeNet` in this test because `execute` passes parameters through CqlTester#transformValues(), which calls
+        // AbstractType#decompose() on the value, which "fixes" the map order, but wouldn't happen normally.
+        executeNet("INSERT INTO %s (k, c) VALUES ('0', ?)", buildInsertionOrderedMap(uuid1, "0", uuid2, "1"));
+        executeNet("INSERT INTO %s (k, c) VALUES ('0', ?)", buildInsertionOrderedMap(uuid2, "3", uuid1, "4"));
+        beforeAndAfterFlush(() -> {
+            assertRowCountNet(executeNet("SELECT * FROM %s WHERE k='0' AND c={" + uuid1 + ": '0', " + uuid2 + ": '1'}"), 1);
+            assertRowCountNet(executeNet("SELECT * FROM %s WHERE k='0' AND c={" + uuid2 + ": '1', " + uuid1 + ": '0'}"), 1);
+            assertRowCountNet(executeNet("SELECT * FROM %s WHERE k='0' AND c={" + uuid1 + ": '4', " + uuid2 + ": '3'}"), 1);
+            assertRowCountNet(executeNet("SELECT * FROM %s WHERE k='0' AND c={" + uuid2 + ": '3', " + uuid1 + ": '4'}"), 1);
+        });
+    }
+
+    @Test
+    public void testInsertingMapDataWithParameterizedQueriesIsKeyOrderIndependentWithSelectSlice() throws Throwable
+    {
+        UUID uuid1 = UUIDs.timeBased();
+        UUID uuid2 = UUIDs.timeBased();
+        createTable("CREATE TABLE %s (k text, c frozen<map<timeuuid, text>>, PRIMARY KEY (k, c));");
+        // We use `executeNet` in this test because `execute` passes parameters through CqlTester#transformValues(), which calls
+        // AbstractType#decompose() on the value, which "fixes" the map order, but wouldn't happen normally.
+        beforeAndAfterFlush(() -> {
+            executeNet("INSERT INTO %s (k, c) VALUES ('0', ?)", buildInsertionOrderedMap(uuid2, "2", uuid1, "1"));
+            // This works before the fix because uuid2 was inserted/serialized first
+            assertRowsNet(executeNet("SELECT k, c[" + uuid2 + "] from %s"), row("0", "2"));
+            // This fails because uuid1 is < uuid2, but is is serialized after,
+            // we short-circuit the projection once we think we're past the potential map key.
+            // Therefore, the query ends up returning `null` for the map value
```

Comment:
I would either update or remove the comments of the form "this works"/"this fails". These comments are useful in the context of presenting this bugfix but, once we had fixed the bug, the comments are obsolete because the bug is already fixed and the query doesn't fail anymore.

Reply:
Fair... will remove them.

Reply:
Fair. What I usually do is tell the story. For example...

> Before CASSANDRA-17623, this failed, because uuid1 is < uuid2 but was serialized after...

Reply:
Added a narrative at the top of the new set of tests in CollectionsTest and a marker at the end of those tests specifying they were all for CASSANDRA-17623, which also allowed me to remove some of the duplicated comments in the tests themselves.